src/directory_tree.cpp

```diff
@@ -0,0 +1,291 @@
+				has_fast_dir_stat = false;
+			} else {
+				is_directory = type == Platform::FileType::Directory;
+			}
+		}
+
+		if (!has_fast_dir_stat) {
+			is_directory = FileFinder::IsDirectory(FileFinder::MakePath(full_path, name), true);
+		}
+
+		if (name == "." || name == "..") {
+			continue;
+		}
+
+		if (is_directory) {
+			std::string new_entry_key = make_key(name);
+			if (std::find_if(entries.begin(), entries.end(), [&](const auto& e) {
+				return e.type == DirectoryTree::FileType::Directory && make_key(e.name) == new_entry_key;
+			}) != entries.end()) {
+				Output::Warning("This game provides the folder \"{}\" twice.", name);
+				Output::Warning("This can lead to file not found errors. Merge the directories manually in a file browser.");
+			}
+		}
+
+		entries.emplace_back(
+			name,
+			is_directory ? FileType::Directory : FileType::Regular);
+	}
+
+	dir_cache[dir_key] = fs_path;
+
+	DirectoryListType fs_cache_entry;
+
+	for (auto& entry : entries) {
+		fs_cache_entry.emplace(std::make_pair(make_key(entry.name), entry));
+	}
+	fs_cache.emplace(dir_key, fs_cache_entry);
+
+	dir_it = dir_cache.find(dir_key);
+
+	return &fs_cache.find(dir_key)->second;
+}
+
+DirectoryTreeView DirectoryTree::Subtree(std::string sub_path) {
+	return DirectoryTreeView(this, std::move(sub_path));
+}
+
+std::string DirectoryTree::FindFile(StringView filename, Span<StringView> exts) const {
+	return FindFile({ ToString(filename), exts });
+}
+
+std::string DirectoryTree::FindFile(StringView directory, StringView filename, lcf::Span<StringView> exts) const {
+	return FindFile({ FileFinder::MakePath(directory, filename), exts });
+}
+
+std::string DirectoryTree::FindFile(const DirectoryTree::Args& args) const {
+	std::string dir, name, canonical_path;
+	// Few games (e.g. Yume2kki) use path traversal (..) in the filenames to point
+	// to files outside of the actual directory.
+	canonical_path = FileFinder::MakeCanonical(args.path, args.canonical_initial_deepness);
+
+#ifdef EMSCRIPTEN
+	if (FileFinder::Exists(canonical_path))
+		return canonical_path;
+#endif
+
+	if (args.translate && !Tr::GetCurrentTranslationId().empty()) {
+		// Search in the active language tree but do not translate again and swallow not found warnings
+		auto translated_file = Tr::GetCurrentTranslationTree().FindFile(
+				{args.path, args.exts, args.canonical_initial_deepness, args.use_rtp, false, false});
+		if (!translated_file.empty()) {
+			DebugLog("Translated {} as {}", args.path, translated_file);
+			return translated_file;
+		}
+	}
+
+	std::tie(dir, name) = FileFinder::GetPathAndFilename(canonical_path);
+	DebugLog("FindFile: {} | {} | {}", canonical_path, dir, name);
+
+	auto* entries = ListDirectory(dir);
+	if (!entries) {
+		if (args.file_not_found_warning) {
+			Output::Debug("Cannot find: {}/{}", dir, name);
+		}
+		return "";
+	}
+
+	std::string dir_key = make_key(dir);
+	auto dir_it = dir_cache.find(dir_key);
+	assert(dir_it != dir_cache.end());
```

Comment:
`dir_it` not used except for in the assert, I would wrap it all in `NDEBUG` so the `dir_cache` lookup isn't done in release build